crates/core/src/factories/core_fetch/test/response_queue.rs

```diff
@@ -0,0 +1,228 @@
+        url: Some(Some(Url::from_str("wss://127.0.0.1:1").unwrap())),
+        ..Default::default()
+    }
+    .build();
+    let agent_url_1 = agent_info_1.url.clone().unwrap();
+    peer_store.insert(vec![agent_info_1.clone()]).await.unwrap();
+
+    let op_id_3 = random_op_id();
+    let op_id_4 = random_op_id();
+    let agent_id_2 = random_agent_id();
+    let agent_info_2 = AgentBuilder {
+        agent: Some(agent_id_2.clone()),
+        url: Some(Some(Url::from_str("wss://127.0.0.1:2").unwrap())),
+        ..Default::default()
+    }
+    .build();
+    let agent_url_2 = agent_info_2.url.clone().unwrap();
+    peer_store.insert(vec![agent_info_2.clone()]).await.unwrap();
+
+    // Insert ops to be read and sent into op store.
+    let op_1 = MetaOp {
+        op_id: op_id_1.clone(),
+        op_data: serde_json::to_vec(&Kitsune2MemoryOp::new(
+            op_id_1.clone(),
+            Timestamp::now(),
+            vec![1; 128],
+        ))
+        .unwrap(),
+    };
+    let op_2 = MetaOp {
+        op_id: op_id_2.clone(),
+        op_data: serde_json::to_vec(&Kitsune2MemoryOp::new(
+            op_id_2.clone(),
+            Timestamp::now(),
+            vec![2; 128],
+        ))
+        .unwrap(),
+    };
+    let op_3 = MetaOp {
+        op_id: op_id_3.clone(),
+        op_data: serde_json::to_vec(&Kitsune2MemoryOp::new(
+            op_id_3.clone(),
+            Timestamp::now(),
+            vec![3; 128],
+        ))
+        .unwrap(),
+    };
+    // Insert op 1, 2, 3 into op store. Op 4 will not be returned in the response.
+    let ops_to_store = vec![op_1.clone(), op_2.clone(), op_3.clone()];
+    op_store.process_incoming_ops(ops_to_store).await.unwrap();
+
+    let fetch = CoreFetch::new(
+        config.clone(),
+        space_id.clone(),
+        peer_store.clone(),
+        op_store,
+        mock_transport.clone(),
+    );
+
+    // Receive 2 op requests.
+    let requested_ops_1 = vec![op_id_1.clone(), op_id_2.clone()];
+    let requested_ops_2 = vec![op_id_3.clone(), op_id_4.clone()];
+    futures::future::join_all([
+        fetch.respond_with_ops(requested_ops_1, agent_id_1.clone()),
+        fetch.respond_with_ops(requested_ops_2, agent_id_2.clone()),
+    ])
+    .await;
+
+    tokio::time::timeout(Duration::from_millis(10), async {
+        loop {
+            tokio::time::sleep(Duration::from_millis(1)).await;
+            if mock_transport.responses_sent.lock().unwrap().len() == 2 {
+                break;
+            }
+        }
+    })
+    .await
+    .unwrap();
+
+    assert!(mock_transport
+        .responses_sent
+        .lock()
+        .unwrap()
+        .contains(&(vec![op_1, op_2], agent_url_1)));
+    // Only op 3 is in op store.
+    assert!(mock_transport
+        .responses_sent
+        .lock()
+        .unwrap()
+        .contains(&(vec![op_3], agent_url_2)));
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn no_response_sent_when_no_ops_found() {
```

Comment:
Possible to also test agent info not in peer store?

Reply:
and also it'd be great to test error, and then queuing the request again for success. Just to demonstrate that there's no state preventing that working?

Reply:
Sure, tested those two aspects https://github.com/holochain/kitsune2/pull/63/commits/a3ee21fd8b6547405a5cbdf5ae4ca34282e84b3e

(Look at the commit while I haven't rewritten git history yet ;-)

Reply:
I've taken out the test for agent info in peer store, because the response is now sent directly to the incoming url.